In [3]:
using Pkg
Pkg.add("Latexify");
Pkg.add("LaTeXStrings");
Pkg.add("LinearAlgebra");
Pkg.add("JuMP");
Pkg.add("Clp");
Pkg.add("NamedArrays")

  Resolving package versions...
No Changes to `~/.julia/environments/v1.5/Project.toml`
No Changes to `~/.julia/environments/v1.5/Manifest.toml`
  Resolving package versions...
No Changes to `~/.julia/environments/v1.5/Project.toml`
No Changes to `~/.julia/environments/v1.5/Manifest.toml`
  Resolving package versions...
No Changes to `~/.julia/environments/v1.5/Project.toml`
No Changes to `~/.julia/environments/v1.5/Manifest.toml`
  Resolving package versions...
No Changes to `~/.julia/environments/v1.5/Project.toml`
No Changes to `~/.julia/environments/v1.5/Manifest.toml`
  Resolving package versions...
No Changes to `~/.julia/environments/v1.5/Project.toml`
No Changes to `~/.julia/environments/v1.5/Manifest.toml`
  Resolving package versions...
No Changes to `~/.julia/environments/v1.5/Project.toml`
No Changes to `~/.julia/environments/v1.5/Manifest.toml`


In [2]:
println("Homework 3 - Jacob Lorenz")

Homework 3 - Jacob Lorenz


# Problem 1 - Doodle Interview Scheduling

**Doodle inc is looking to interview a candidate for a new software engineer position at their company. It works like this: The interview (10AM - 3PM) is divided into anumber of 20-minute time slots that may be used for 1-on-1 meetings with the candidate. There is lso a one-hour time slot in the middle of the day where 3 employees take the candidate out for lunch.**

**It would be nice for all 15 senior employees to meet with the candidate at some point during the day, but everybody has a busy schedule so it's not clear whether this will be possible. A doodle poll was sent to the 15 senior employees to figure out their availability. See hw3data.ipynb for the incidence matrix describing the availabilities of each employee at the various time slots.**

**In the table, a 1 means that the employee is available at the indicated time, while a 0 means that they are unavailable. Determine whether a feasible interview schedule exists. If so, print out a calendar for the candidate that lists who they will be meeting at each time slot.**

When we take some time to think through this problem and process the important information and objectives, we can see this problem effectively becomes a transportation problem. In this case, we can reduce this problem to a more generalized template. We notice that we can represent each employee as a source node since the employee is providing the availability, and represent each time slot as a sink node demanding availability. We can then place constraints on the edges to represent things such as capacity and availabilitiy restrictions. Finally, this problem is a maximization problem and each cost coefficient will just be 1, thus the LP can be thought of as wanting to maximize the number of edges that have capacity on them in order to fill as many time slots as possible. Thus, we formulate our model as follows:

Let $x_{ij}$ represent the flow of time-availability going from employee i to project j where $i \in [1,15]$ and $j \in [a,m]$ (13 time slots). We can apply constraints to these flow variables to model the particularities of our system. We will cover these constraints in 3 parts, supply, capacity, and demand.

$\textbf{Supply:}$

For supply, we know that each employee can only fill one time slot, so the sum of all flow outgoing from that employee's source node must sum to one with no negative flow. Although we would like for all of these to be met with equality, if there is a feasible solution, the solver will maximize it, so we can instead save ourself half of the constraints by using the inequality constraint. So, we can model that as follows, remembering our convention for the flow variable denotations:

$$
    x_{1a} + x_{1b} + x_{1c} + x_{1d} + x_{1e} + x_{1f} + x_{1g} + x_{1h} + x_{1i} + x_{1j} + x_{1k} + x_{1l} + x_{1m} \leq 1 \\
    x_{2a} + x_{2b} + x_{2c} + x_{2d} + x_{2e} + x_{2f} + x_{2g} + x_{2h} + x_{2i} + x_{2j} + x_{2k} + x_{2l} + x_{2m} \leq 1 \\
    x_{3a} + x_{3b} + x_{3c} + x_{3d} + x_{3e} + x_{3f} + x_{3g} + x_{3h} + x_{3i} + x_{3j} + x_{3k} + x_{3l} + x_{3m} \leq 1 \\
    x_{4a} + x_{4b} + x_{4c} + x_{4d} + x_{4e} + x_{4f} + x_{4g} + x_{4h} + x_{4i} + x_{4j} + x_{4k} + x_{4l} + x_{4m} \leq 1 \\
    x_{5a} + x_{5b} + x_{5c} + x_{5d} + x_{5e} + x_{5f} + x_{5g} + x_{5h} + x_{5i} + x_{5j} + x_{5k} + x_{5l} + x_{5m} \leq 1 \\
    x_{6a} + x_{6b} + x_{6c} + x_{6d} + x_{6e} + x_{6f} + x_{6g} + x_{6h} + x_{6i} + x_{6j} + x_{6k} + x_{6l} + x_{6m} \leq 1 \\
    x_{7a} + x_{7b} + x_{7c} + x_{7d} + x_{7e} + x_{7f} + x_{7g} + x_{7h} + x_{7i} + x_{7j} + x_{7k} + x_{7l} + x_{7m} \leq 1 \\
    x_{8a} + x_{8b} + x_{8c} + x_{8d} + x_{8e} + x_{8f} + x_{8g} + x_{8h} + x_{8i} + x_{8j} + x_{8k} + x_{8l} + x_{8m} \leq 1 \\
    x_{9a} + x_{9b} + x_{9c} + x_{9d} + x_{9e} + x_{9f} + x_{9g} + x_{9h} + x_{9i} + x_{9j} + x_{9k} + x_{9l} + x_{9m} \leq 1 \\
    x_{10a} + x_{10b} + x_{10c} + x_{10d} + x_{10e} + x_{10f} + x_{10g} + x_{10h} + x_{10i} + x_{10j} + x_{10k} + x_{10l} + x_{10m} \leq 1 \\
    x_{11a} + x_{11b} + x_{11c} + x_{11d} + x_{11e} + x_{11f} + x_{11g} + x_{11h} + x_{11i} + x_{11j} + x_{11k} + x_{11l} + x_{11m} \leq 1 \\
    x_{12a} + x_{12b} + x_{12c} + x_{12d} + x_{12e} + x_{12f} + x_{12g} + x_{12h} + x_{12i} + x_{12j} + x_{12k} + x_{12l} + x_{12m} \leq 1 \\
    x_{13a} + x_{13b} + x_{13c} + x_{13d} + x_{13e} + x_{13f} + x_{13g} + x_{13h} + x_{13i} + x_{13j} + x_{13k} + x_{13l} + x_{13m} \leq 1 \\
    x_{14a} + x_{14b} + x_{14c} + x_{14d} + x_{14e} + x_{14f} + x_{14g} + x_{14h} + x_{14i} + x_{14j} + x_{14k} + x_{14l} + x_{14m} \leq 1 \\
    x_{15a} + x_{15b} + x_{15c} + x_{15d} + x_{15e} + x_{15f} + x_{15g} + x_{15h} + x_{15i} + x_{15j} + x_{15k} + x_{15l} + x_{15m} \leq 1
$$

$\textbf{Demand:}$

Demand can be thought of as the number of employees required and limited to fill each time slot. Again, to make things simpler we will formulate these as $\leq$ inequalities even though we would like for them to all have equality, since we know that the solver will maximize the total pairings if feasible. Broadly, most of the time slots have a limit of 1, except for the lunch spot, which has a demand of 3.

$$
    x_{1a} + x_{2a} + x_{3a} + x_{4a} + x_{5a} + x_{6a} + x_{7a} + x_{8a} + x_{9a} + x_{10a} + x_{11a} + x_{12a} + x_{13a} + x_{14a} + x_{15a} \leq 1 \\
    x_{1b} + x_{2b} + x_{3b} + x_{4b} + x_{5b} + x_{6b} + x_{7b} + x_{8b} + x_{9b} + x_{10b} + x_{11b} + x_{12b} + x_{13b} + x_{14b} + x_{15b} \leq 1 \\
    x_{1c} + x_{2c} + x_{3c} + x_{4c} + x_{5c} + x_{6c} + x_{7c} + x_{8c} + x_{9c} + x_{10c} + x_{11c} + x_{12c} + x_{13c} + x_{14c} + x_{15c} \leq 1 \\
    x_{1d} + x_{2d} + x_{3d} + x_{4d} + x_{5d} + x_{6d} + x_{7d} + x_{8d} + x_{9d} + x_{10d} + x_{11d} + x_{12d} + x_{13d} + x_{14d} + x_{15d} \leq 1 \\
    x_{1e} + x_{2e} + x_{3e} + x_{4e} + x_{5e} + x_{6e} + x_{7e} + x_{8e} + x_{9e} + x_{10e} + x_{11e} + x_{12e} + x_{13e} + x_{14e} + x_{15e} \leq 1 \\
    x_{1f} + x_{2f} + x_{3f} + x_{4f} + x_{5f} + x_{6f} + x_{7f} + x_{8f} + x_{9f} + x_{10f} + x_{11f} + x_{12f} + x_{13f} + x_{14f} + x_{15f} \leq 1 \\
    x_{1g} + x_{2g} + x_{3g} + x_{4g} + x_{5g} + x_{6g} + x_{7g} + x_{8g} + x_{9g} + x_{10g} + x_{11g} + x_{12g} + x_{13g} + x_{14g} + x_{15g} \leq 3 \\
    x_{1h} + x_{2h} + x_{3h} + x_{4h} + x_{5h} + x_{6h} + x_{7h} + x_{8h} + x_{9h} + x_{10h} + x_{11h} + x_{12h} + x_{13h} + x_{14h} + x_{15h} \leq 1 \\
    x_{1i} + x_{2i} + x_{3i} + x_{4i} + x_{5i} + x_{6i} + x_{7i} + x_{8i} + x_{9i} + x_{10i} + x_{11i} + x_{12i} + x_{13i} + x_{14i} + x_{15i} \leq 1 \\
    x_{1j} + x_{2j} + x_{3j} + x_{4j} + x_{5j} + x_{6j} + x_{7j} + x_{8j} + x_{9j} + x_{10j} + x_{11j} + x_{12j} + x_{13j} + x_{14j} + x_{15j} \leq 1 \\
    x_{1k} + x_{2k} + x_{3k} + x_{4k} + x_{5k} + x_{6k} + x_{7k} + x_{8k} + x_{9k} + x_{10k} + x_{11k} + x_{12k} + x_{13k} + x_{14k} + x_{15k} \leq 1 \\
    x_{1l} + x_{2l} + x_{3l} + x_{4l} + x_{5l} + x_{6l} + x_{7l} + x_{8l} + x_{9l} + x_{10l} + x_{11l} + x_{12l} + x_{13l} + x_{14l} + x_{15l} \leq 1 \\
    x_{1m} + x_{2m} + x_{3m} + x_{4m} + x_{5m} + x_{6m} + x_{7m} + x_{8m} + x_{9m} + x_{10m} + x_{11m} + x_{12m} + x_{13m} + x_{14m} + x_{15m} \leq 1 \\
$$

$\textbf{Capacity:}$

The capacity constraints we put on our system are what give rise to the system's characteristics. This is where we can place restrictions on which employees are available to fill particular time slots (we can set the flow restriction on variables such as $x_{1a}$ to 0 to simulate Employee 1 not having any availability for the first 20-minute meeting spot). Since there are 15 * 13 possible edges, I am not going to list them all here, rather, my approach is to declare all variables as $\geq 0$ and $\leq n$ where n represents the binary value indicating whether the respective employee is available for a particular time slot. For example,

$$
    0 \leq x_{1a} \leq 0 \\
    0 \leq x_{1b} \leq 0 \\
    0 \leq x_{1c} \leq 1 \\
    0 \leq x_{1d} \leq 1 \\
    0 \leq x_{1e} \leq 0 \\
    0 \leq x_{1f} \leq 0 \\
    0 \leq x_{1g} \leq 0 \\
    0 \leq x_{1h} \leq 1 \\
    0 \leq x_{1i} \leq 1 \\
    0 \leq x_{1j} \leq 0 \\
    0 \leq x_{1k} \leq 0 \\
    0 \leq x_{1l} \leq 0 \\
    0 \leq x_{1m} \leq 0 \\
    \dots
$$

Given the fact that the incidence matrix for my formulation, A, has 223 rows and 195 columns, I am not going to write it out here, nor am I going to explicitely type out the 195-element columns vector b or 195-element columsn vector c, which consists of all ones. We will let x be translated to the following for the sake of indexing in the model:

$$
    \text{x} = 
    \left[ \begin{array}{rl}
                x_{1a} \\
                x_{1b} \\
                x_{1c} \\
                x_{1d} \\
                x_{1e} \\
                \dots \\
                x_{15h} \\
                x_{15i} \\
                x_{15j} \\
                x_{15k} \\
                x_{15l} \\
                x_{15m} \\
            \end{array}\right]
            = \left[ \begin{array}{rl}
                x_{1} \\
                x_{2} \\
                x_{3} \\
                x_{4} \\
                x_{5} \\
                \dots \\
                x_{190} \\
                x_{191} \\
                x_{192} \\
                x_{193} \\
                x_{194} \\
                x_{195} \\
            \end{array}\right]
$$

The objective of our model is the following:

$$
    \text{max} \;\; c^Tx \\
$$

Finally, let's solve it via dynamic constraint calculation:

In [15]:
using JuMP
using Clp
using NamedArrays


availability =
  [ 0 0 1 1 0 0 0 1 1 0 0 0 0
    0 1 1 0 0 0 0 0 1 1 0 0 0
    0 0 0 1 1 0 1 1 0 1 1 1 1
    0 0 0 1 1 1 1 1 1 1 1 1 0
    0 0 0 0 0 0 1 1 1 0 0 0 0
    0 1 1 0 0 0 0 0 1 1 0 0 0
    0 0 0 1 1 1 1 0 0 0 0 0 0
    1 1 0 0 0 0 0 0 0 0 1 1 1
    1 1 1 0 0 0 0 0 0 1 1 0 0
    0 0 0 0 0 0 0 1 1 0 0 0 0
    0 0 0 0 0 0 1 1 1 0 0 0 0
    1 1 0 0 0 1 1 1 1 0 0 1 1
    1 1 1 0 1 1 0 0 0 0 0 1 1
    0 1 1 1 0 0 0 0 0 0 0 0 0
    1 1 0 0 1 1 0 0 0 0 0 0 0 ]

TIMES = ["10:00","10:20","10:40","11:00","11:20","11:40","lunch","1:00","1:20","1:40","2:00","2:20","2:40"]
NAMES = [:Manuel,:Luca,:Jule,:Michael,:Malte,:Chris,:Spyros,:Mirjam,:Matt,:Florian,:Josep,:Joel,:Tom,:Daniel,:Anne ]
times = NamedArray( availability, (NAMES,TIMES), ("NAME","TIME"))

NumEmployees = size(NAMES,1); # 15
NumTimeslots = size(TIMES,1); #13

NumberEdges = NumEmployees * NumTimeslots;

x = [j for j in 1:NumberEdges];
c = ones(NumberEdges);

m1 = Model();

@variable(m1, x[x] >= 0)

# Supply constraints
for i in 1:NumEmployees
    @constraint(m1, sum(x[13 * (i-1) + j] for j in 1:NumTimeslots) <= 1)
end

# Demand Constraints
for j in 1:NumTimeslots
    if (j == 7)
        @constraint(m1,sum(x[13 * (i-1) + j] for i in 1:NumEmployees) <= 3)
    else
        @constraint(m1,sum(x[13 * (i-1) + j] for i in 1:NumEmployees) <= 1)
    end
end

# Capacity Constraints
for i in 1:NumEmployees
   for j in 1:NumTimeslots 
        @constraint(m1,x[13 * (i-1) + j] <= availability[i,j])
    end
end

# Define the objective - we want to maximize here
@objective(m1, Max, sum(c[i] * x[i] for i in 1:NumberEdges));

println("Stats for the Clp Solver")
println("===============================")

# Set your chosen optimizer and solve with stat output
set_optimizer(m1, Clp.Optimizer)
@time optimize!(m1)

Stats for the Clp Solver
 11.261127 seconds (42.43 M allocations: 2.120 GiB, 8.62% gc time)
Coin0506I Presolve 28 (-195) rows, 73 (-122) columns and 146 (-439) elements
Clp0006I 0  Obj -0 Dual inf 72.999927 (73)
Clp0006I 33  Obj 15 Primal inf 69.799977 (23)
Clp0006I 64  Obj 15 Primal inf 18.19998 (20)
Clp0006I 85  Obj 15
Clp0000I Optimal - objective value 15
Coin0511I After Postsolve, objective 15, infeasibilities - dual 0 (0), primal 0 (0)
Clp0032I Optimal objective 15 - 85 iterations time 0.002, Presolve 0.00


In [65]:
# Print out the optimal values for each model variable
timeslots = Dict{Int32,Set}(i => Set() for i = 1:NumTimeslots)
for i in 1:NumberEdges
    if (value(x[i]) > 0)
        # Get the indices for employee and time
        timeslotIndex = ((i - 1) % 13) + 1;
        employeeIndex = Int(((i - timeslotIndex) / 13) + 1);
        # Add the index of the employee in the set corresponding with this time slot index
        push!(timeslots[timeslotIndex], employeeIndex)
    end
end

# Go through and print out the schedule
println("Schedule:")
println("==========================")
for i in 1:NumTimeslots
    print(TIMES[i])
    print(": ")
    employeeIds = timeslots[i];
    if length(employeeIds) > 1
        for i in employeeIds
           print(NAMES[i], " "); 
        end
        println()
    else
        for i in employeeIds
           print(NAMES[i]); 
        end
        println()
    end
end
println()

Schedule:
10:00: Mirjam
10:20: Luca
10:40: Manuel
11:00: Daniel
11:20: Jule
11:40: Anne
lunch: Spyros Malte Joel 
1:00: Florian
1:20: Josep
1:40: Chris
2:00: Matt
2:20: Michael
2:40: Tom



In conclusion of problem 1, we have determined that a feasible solution does indeed exist such that each of the 15 employees is able to meet with the candidate for one time slot or another, and 3 that take the candidate out to lunch. The schedule can be found in the results in the cell immediately above.

# Problem 2 - Car Rental

**A small car rental company has a fleet of 94 vehicles distributed among its 10 agencies. The location of every agency is given by its geographical coordinates x and y in a grid based on miles. We assume that the road distance between agencies is approximately 1.3 times the Euclidean distance (as the crow flies). The table in HW Problem set 3 indicates the coordinates of all agencies, the number of cars required the next morning, and the stock of cars in the evening preceding this day.**

**Supposing the cost for transporting a car is 50 cents per mile, determine the movements of cars that allow the company to re-establish the required number of cars at all agencies, minimizing the total cost incurred for transport**

This was a fun problem to solve. I began a little unsure of how to approach it. First thing I did was get all the data in a consistently indexed and organized format. After this, I thought of what the decision variables in the problem are. After some thinking, we see that this problem reduces to a transportation problem with a slight twist. We can think of each agency as having both supply and demand, and the decision variables, which represent the flow of cars from one agency to another, indicate the number of vehicles one agency should send to another specific agency.

Formally, $x_{ij}$ represents the number of cars that agency $i$ transferred to agency $j$ for  $i \in [1,10]$ and $j \in [1,10]$. For the sake of naming convention and eliminating confusion, we will let the $i$ index be represented numerically, and we will convert the $j$ index to an equivalent alphabetical range, $j \in [a,j]$. Then, we have

$$
    \text{x} = 
    \left[ \begin{array}{rl}
                x_{1a} \\
                x_{1b} \\
                x_{1c} \\
                x_{1d} \\
                x_{1e} \\
                \dots \\
                x_{10f} \\
                x_{10g} \\
                x_{10h} \\
                x_{10i} \\
                x_{10j} \\
            \end{array}\right]
            = \left[ \begin{array}{rl}
                x_{1} \\
                x_{2} \\
                x_{3} \\
                x_{4} \\
                x_{5} \\
                \dots \\
                x_{96} \\
                x_{97} \\
                x_{98} \\
                x_{99} \\
                x_{100} \\
            \end{array}\right]
$$

The next thing to address is the coordinates of each agency and the location relative to the other agencies. The distance between each agency is directly relevant to the cost function, so we need to calculate the Euclidean Distance between each pair of agencies. For agencies a and b, the Euclidean Distance in terms of x and y coordinates can be expressed as

$$
    ED_{a,b} = \sqrt{(a_x - b_x)^2 + (a_y - b_y)^2}
$$

Furthermore, we we get our cost matrix by multiplying the Euclidean Distance matrix by 1.3, and then again by 0.5 to represent the difference in road distance and the cost per mile. One thing to note here is that our x variables are sequential from 1 to 100; however, our cost matrix is a 10x10 matrix that will not be able to be directly multiplied. We will not go into the details of indexing here, but this is circumvented by some index decomposition to translate 0 to 100 into two indices between 0 and 10.

$$
    [Cost] = 0.5 * 1.3 * [ED]
$$

With that, we now have our decision variables and our cost function, the remaining item is to describe the constraints. Typically I break problems down into 3 subsets of constraints, supply, demand, and capacity. I typically think of capacity constraints as being particularities of the system to set time restrictions or maximum workload; however, in this model there are actually no capacity constraints, only supply and demand.

**Supply:**

As previously mentioned, each agency acts as both a producer and consumer of flow in this model. As a supplier, each agency is only capable of sending a number of cars they currently have, which is specified in the CurrentInventory column vector. So, any and all agencies a particular agency sends cars to, the quantity of cars cannot exceed their current inventory. I am not going to type out dozens of linear combinations, instead, I will write out the formal mathematical expression to describe this section of constraints. In total, there will be 10 supply constraints.

$$
    \sum_{j = 1}^{10}{x[10 * (i-1) + j]} \leq CurrentInventory_i \;\;\; \forall i \in [1,10]
$$

**Demand:**

The demand constraints are similar to the supply constraints, but with a twist. Each agency has the potential to receive cars from any other agency. Each agency can send cars. Each agency starts with an initial amount of cars. The demand constraints for an agency work to enforce a certain number of net cars to end up at any given agency. So the net flow in conjunction with initial values needs to be above a lower bound for each agency. Let CurrentInventory be the column vector specifying the number of cars each agency starts with, and NumberRequiredCars be the column vector specifying the number of cars each agency needs to end up with. Then we have the following set of demand constraints:

$$
    CurrentInventory_i + \sum_{j = 1}^{10}{x[10 * (j-1) + i]} - \sum_{j = 1}^{10}{x[10 * (i-1) + j]} \geq NumberRequiredCars_i \;\;\; \forall i \in [1,10]
$$

Finally, the objective is to minimize the overall cost, which can be found with a linear combination of the x vector and Cost matrix. The indexing was interesting to figure out for this one. Understanding that the indices of x go from 1 to 100, and the indices of the x and y components of the Cost matrix go from 1 to 10, and also heeding what each cell in the matrix represents, we have the following transoformation from the scale of 1 to 100 to a pair of indices between 1 and 10. Let i represent the index between 1 and 100. Then total cost is given by

$$
    \sum_{i=1}^{100}{x[i] * Costs[((i - 1) \; \% \; 10) + 1][((i - (((i - 1) \; \% \; 10) + 1)) \; / \; 10) + 1]}
$$

We want to minimize this value. With all that being said, we have described the formulation of our model as succinctly as we can, and now we solve:

In [120]:
# Put the data into usable and consistent matrices
##################################################
Agencies = [
    1; 2; 3; 4; 5; 6; 7; 8; 9; 10
];

NumberAgencies = size(Agencies,1);

NumberRequiredCars = [
    10; 6; 8; 11; 9; 7; 15; 7; 9; 12
];

CurrentInventory = [
    8
    13
    4
    8
    12
    2
    14
    11
    15
    7
];

AgencyCoordinates = [
    0 0
    20 20
    18 10
    30 12
    35 0
    33 25
    5 27
    5 10
    11 0
    2 15
];

EuclideanDistances = zeros(NumberAgencies,NumberAgencies);

# Calculate the Euclidean Distance between each pair of Agencies
for i in 1:NumberAgencies
    for j in 1:NumberAgencies
        I_X_Coord = AgencyCoordinates[i,1];
        I_Y_Coord = AgencyCoordinates[i,2];
        
        J_X_Coord = AgencyCoordinates[j,1];
        J_Y_Coord = AgencyCoordinates[j,2];
        
        # Calculate the euclidean distance
        curDistance = sqrt((I_X_Coord - J_X_Coord)^2 + (I_Y_Coord - J_Y_Coord)^2);
        EuclideanDistances[i,j] = curDistance;
    end
end

Costs = 0.5 * 1.3 * EuclideanDistances;

##################################################

# We now have a matrix containing the Euclidean Distances between each agency. It's 
# important to remember that these mileage values need to be 1.3x for road mileage.
# To do:
# - Supply Constraints
# - Demand Constraints
# - Cost Function
# - Solve


x = [i for i in 1:NumberAgencies^2];

m2 = Model();

# Create variables for each x in the values specified above
# Some manipulation will be needed later, 1a, 1b, 1c, ... -> 10j = 1, 2, 3, ... -> 100
@variable(m2, x[x] >= 0)

# Supply Constraints
for i in 1:NumberAgencies
    @constraint(m2, sum(x[10 * (i-1) + j] for j in 1:NumberAgencies) <= CurrentInventory[i])
end

# Demand Constraints
for i in 1:NumberAgencies
    InitialSupply = CurrentInventory[i];
    LostSupply = sum(x[10 * (i-1) + j] for j in 1:NumberAgencies);
    GainedSupply = sum(x[10 * (j-1) + i] for j in 1:NumberAgencies);
    @constraint(m2, InitialSupply + GainedSupply - LostSupply >= NumberRequiredCars[i])
end

# Cost Function - Define the objective - we want to minimize here
@objective(m2, Min, sum(Costs[Int(((i-1) % 10) + 1), Int(((i - (((i-1) % 10) + 1)) / 10) + 1)] * x[i] for i in 1:NumberAgencies^2))

println("Stats for the Clp Solver")
println("===============================")

# Set your chosen optimizer and solve with stat output
set_optimizer(m2, Clp.Optimizer)
@time optimize!(m2)

# Print out the values that are > 0
for i in 1:NumberAgencies^2
    if (value(x[i]) > 0)
        println("Value for x[",i,"]: ", value(x[i]));
    end
end

Stats for the Clp Solver
  0.001887 seconds (1.81 k allocations: 223.172 KiB)
Value for x[13]: 1.0
Value for x[16]: 5.0
Value for x[17]: 1.0
Value for x[44]: 3.0
Value for x[80]: 4.0
Value for x[81]: 2.0
Value for x[83]: 3.0
Value for x[90]: 1.0
Coin0506I Presolve 20 (0) rows, 90 (-10) columns and 270 (-10) elements
Clp0006I 0  Obj 0 Primal inf 19.999994 (6)
Clp0006I 11  Obj 152.63902
Clp0000I Optimal - objective value 152.63902
Coin0511I After Postsolve, objective 152.63902, infeasibilities - dual 0 (0), primal 0 (0)
Clp0032I Optimal objective 152.6390163 - 11 iterations time 0.002, Presolve 0.00


In conclusion of problem 2, we see in the results above that the optimal transfer of cars to minimize cost is as follows:
- Agency 2 transfers 1 car to Agency 3
- Agency 2 transfers 5 cars to Agency 6
- Agency 2 transfers 1 car to Agency 7
- Agency 5 transfers 3 crs to Agency 4
- Agency 8 transfers 4 cars to Agency 10
- Agency 9 transfers 2 cars to Agency 1
- Agency 9 transfers 3 cars to Agency 3
- Agency 9 transfers 1 car to Agency 10

If you take the time to go through understand the impact of each of these moves, you will see that each Agency does indeed end up with the required number of cars. Last but not least, the optimal transfer of all these vehicles will cost $\$152.64$ if they are paying 50 cents per car per mile.

# Problem 3 - Building a Stadium

**A town council wishes to construct a small stadium in order to improve the services provided to the people living in the district. After the invitation to tender, a local construction company is awarded the contract and wishes to complete the task within the shortest possible time. All the major tasks are listed in the table provided in hw3data.ipynb. Some tasks can only start after the completion of certain other tasks, as indicated by the "predecessors" column.**

**a) What is the earliest possible date of completion for the construction? Note that the last two columns of the table are not relevant for this part of the problem.**

First we will load in all the data for our program.

In [184]:
Tasks = [i for i in 1:19];
Durations = [2, 16, 9, 8, 10, 6, 2, 2, 9, 5, 3, 2, 1, 7, 4, 3, 9, 1, 0];
DurationsDict = Dict(zip(Tasks, Durations));
Predecessors = [ [], [1], [2], [2], [3], [4,5], [4], [6], [4,6], [4], [6], [9], [7], [2], [4,14], [8,11,14], [12], [17], [18] ];
PredecessorsDict = Dict(zip(Tasks, Predecessors));

Next, we will define the mathematical model of our system. 

The first thing to notice is that we added an additional task (19) to the list with a duration of 0 and a dependency on task 18. This task will be used throughout our solution and model to represent the "finish" task, and our goal will revolve around ensuring this finish task is minimized.

**Decision Variables:** In this case, the decisions come down to when to begin each of the 19 tasks while abiding by the predecessor requirements and understanding the task durations. Let $t_i$ represent the time at which task $i$ is started $\forall \; i \in \tau$ where $\tau$ is the set of all tasks.

**Constraints:** The constraints are defined by the predecessor listings. A task cannot be started until each of its predecessors have been completed, therefore:

$$
    t_i \geq t_j + d_j\quad \forall j \in \mathcal{P}_i \quad \forall i \in\tau
$$

where $\mathcal{P}_i$ is the set of all predecessor tasks of task $i$ that must be finished before task $i$ can start and $d_j$ is the duration of task $j$.

The first task will begin at time $t_1 == 0$.

**Objective:**

$$
    \min_{t} t_{19}
$$

where $t_{19}$ is the start time and of the finish task. We are now set to solve.

In [185]:
m3 = Model()

@variable(m3, tstart[Tasks])

# one-line implementation of the constraints:
# @constraint(m, link[i in tasks, j in pred[i]], tstart[i] >= tstart[j] + duration[j])

for i in Tasks
    for j in Predecessors[i]
        @constraint(m3, tstart[i] >= tstart[j] + Durations[j])
    end
end
@constraint(m3, tstart[1] == 0)
@objective(m3, Min, tstart[19])     # total duration is start time of last task + duration of last task.

# Set your chosen optimizer and solve with stat output
set_optimizer(m3, Clp.Optimizer)
@time optimize!(m3)
println(value.(tstart))
println("Minimum duration: ", objective_value(m3), " weeks")

  0.326671 seconds (362.31 k allocations: 19.216 MiB, 6.59% gc time)
1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
And data, a 19-element Array{Float64,1}:
  0.0
  2.0
 18.0
 18.0
 27.0
 37.0
 26.0
 43.0
 43.0
 26.0
 43.0
 52.0
 28.0
 18.0
 26.0
 46.0
 54.0
 63.0
 64.0
Minimum duration: 64.0 weeks
Coin0506I Presolve 0 (-24) rows, 0 (-19) columns and 0 (-47) elements
Clp3002W Empty problem - 0 rows, 0 columns and 0 elements
Clp0000I Optimal - objective value 64
Coin0511I After Postsolve, objective 64, infeasibilities - dual 0 (0), primal 0 (0)
Clp0032I Optimal objective 64 - 0 iterations time 0.002, Presolve 0.00


Based on the results obtained above, the minimum duration it will take to build the stadium, Start Installing the Construction Site - Finish Handing over the Building, is **64 weeks**.

**b) The town council wants the builder to expedite the project. As an incentive, the council will pay a bonus of 30k per week for each week the work finishes early. To accomplish this, the builder may employ additional workers and rent more equipment to cut down on the total time. the last two columns of the table show the maximum number of weeks that can be saved per task and the associated cost per week incurred by the extra work. When will the project be completed if the builder is acting in a way that maximizes profit?**

In [181]:
# additional columns of data (maximum reduction possible )
MaxReduction =  [0,  3,  1,  2,  2,  1, 1, 0,  2,  1,  1, 0, 0,  2,  2, 1,  3, 0];  # max reduction (in weeks)
AddtlWeeklyCost = [0, 30, 26, 12, 17, 15, 8, 0, 42, 21, 18, 0, 0, 22, 12, 6, 16, 0];  # cost of reduction ($1,000/week)
WeeklyBonus = 30;     # bonus for expediting the project ($1,000/week )

First, we will define the mathematical model of our system. 

**Decision Variables:** In this case, the first set of decisions comes down to when to begin each of the 19 tasks while abiding by the predecessor requirements and understanding the task durations. Let $t_i$ represent the time at which task $i$ is started $\forall \; i \in \tau$ where $\tau$ is the set of all tasks. Furthermore, we now have another set of decision variables that represent the number of weeks the builder should expedite a particular task. Let $e_i$ represent the number of weeks the builder should expedite task $i$ $\forall \; i \in \tau: i \neq 19$.

**Constraints:** The constraints are first and foremost defined by the predecessor listings and associated task durations. A task cannot be started until each of its predecessors have been completed. Now; however, we have the added decision of reducing any one or more of the predecessor task durations, and that decision must be captured in the constraints. Therefore:

$$
    t_i \geq t_j + (d_j - e_j)\quad \forall j \in \mathcal{P}_i \quad \forall i \in\tau
$$

where $\mathcal{P}_i$ is the set of all predecessor tasks of task $i$ that must be finished before task $i$ can start, $d_j$ is the duration of task $j$, and $e_j$ is the number of weeks that task $j$ was expedited.

Furthermore, we have a limit for the maximum number of weeks a particular task can be expedited as described in the MaxReduction vector, excluding the finish task, specifically:

$$
    0 \leq e_i \leq MaxReduction_i\quad \forall \; i \in \tau: i \neq 19
$$

The first task will begin at time $t_1 == 0$.

Finally, $e_{19}$ will be used to represent the number of weeks that this augmented model finishes ahead of the original model in part a), which if we recall was 64 weeks. We let

$$
    t_{19} = 64 - e_{19}
$$

**Objective:** The objective in this case is to maximize profit, thus, we have

$$
    \max_{t} \text{WeeklyBonus} * e_{19} - \sum_{i=1}^{18}{AddtlWeeklyCost_i * e_i}
$$

where WeeklyBonus is the bonus per week for expediting the project (in $\$1000$), and AddtlWeeklyCost is the column vector describing the cost of expediting each task per week. We are now set to solve.

In [193]:
# additional columns of data (maximum reduction possible )
MaxReduction =  [0,  3,  1,  2,  2,  1, 1, 0,  2,  1,  1, 0, 0,  2,  2, 1,  3, 0];  # max reduction (in weeks)
AddtlWeeklyCost = [0, 30, 26, 12, 17, 15, 8, 0, 42, 21, 18, 0, 0, 22, 12, 6, 16, 0];  # cost of reduction ($1,000/week)
WeeklyBonus = 30;     # bonus for expediting the project ($1,000/week )
OriginalNumberWeeks = 64;
# Tasks = [i for i in 1:19];
# Durations = [2, 16, 9, 8, 10, 6, 2, 2, 9, 5, 3, 2, 1, 7, 4, 3, 9, 1, 0];
# DurationsDict = Dict(zip(Tasks, Durations));
# Predecessors = [ [], [1], [2], [2], [3], [4,5], [4], [6], [4,6], [4], [6], [9], [7], [2], [4,14], [8,11,14], [12], [17], [18] ];
# PredecessorsDict = Dict(zip(Tasks, Predecessors));

m4 = Model()

# Declare decision variables for the start time and the number of expediting weeks
@variable(m4, tstart[Tasks]);
@variable(m4, texpedite[Tasks]);

# Set constraints for the MaxReduction in all but finish task
for i in 1:(size(Tasks,1)-1)
    @constraint(m4, 0 <= texpedite[i] <= MaxReduction[i]);
end

# Set restrictions for predecessors and include the expediting time
for i in Tasks
    for j in Predecessors[i]
        @constraint(m4, tstart[i] >= tstart[j] + Durations[j] - texpedite[j]);
    end
end
# t1 = 0
@constraint(m4, tstart[1] == 0)
# The start time of the finish task should be based on the original and the number of weeks to start the finish task early
@constraint(m4, tstart[size(Tasks,1)] == OriginalNumberWeeks - texpedite[size(Tasks,1)])
# Since 2 is essentially free work / time and is on the immediate critical path,
# we can set this to be equal to its max reduction value
@constraint(m4, texpedite[2] == MaxReduction[2]);

# Our objective is to maximize profit - expenses and remember to take the answer in 1000s
@objective(m4, Max, ((WeeklyBonus * texpedite[size(Tasks,1)]) - sum(texpedite[i] * AddtlWeeklyCost[i] for i in 1:(size(Tasks,1)-1))));     # total duration is start time of last task + duration of last task.

# Set your chosen optimizer and solve with stat output
set_optimizer(m4, Clp.Optimizer)
@time optimize!(m4)
println(value.(tstart))
println(value.(texpedite))

  0.001160 seconds (2.10 k allocations: 191.891 KiB)
1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
And data, a 19-element Array{Float64,1}:
  0.0
  2.0
 15.0
 15.0
 23.0
 31.0
 23.0
 36.0
 36.0
 23.0
 36.0
 45.0
 25.0
 15.0
 23.0
 39.0
 47.0
 53.0
 54.0
1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
And data, a 19-element Array{Float64,1}:
  0.0
  3.0
  1.0
  0.0
  2.0
  1.0
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
  3.0
  0.0
 10.0
Coin0506I Presolve 3 (-41) rows, 7 (-31) columns and 15 (-76) elements
Clp0006I 0  Obj 273 Primal inf 9.0999991 (1)
Clp0006I 1  Obj 87
Clp0000I Optimal - objective value 87
Coin0511I After Postsolve, objective 87, infeasibilities - dual 0 (0), primal 0 (0)
Clp0032I Optimal objective 87 - 1 iterations time 0.002, Presolve 0.00


Based on the results obtained above, the town's incentive would net the builder **$\$$87,000 profit** and would be completed an additional 10 weeks early, bringing the final completion date down from 64 weeks to **54 weeks**.